### PR TITLE
[24464] Finalize FileWriteSchemaTransformProvider

### LINKED ISSUE #24464: [Epic]: Implement FileWriteSchemaTransformProvider
## What work does this Epic describe?

This Epic defines a TypedSchemaTransformProvider extension FileWriteSchemaTransformProvider that builds a transform that encodes Row elements and writes the serialized payload to a file or object system.  The provider unifies and normalizes Beam file writing related details into a single schema aware configuration and produces a file writing transform such as FileIO.Write or TextIO.Write.

## What value may result from this Epic's output?

This Epic adds to the SchemaTransform IO support initiative to enable declarative configuration based IO reads and writes.

### Ready

The following are blockers for beginning this Epic.

- [x] Community review of [design document](https://docs.google.com/document/d/1IOZrQ4qQrUS2WwQhadN35vX4AzhG4dyXMk1J-R1qJ9c/edit?usp=sharing)

### Done

The following are measures for marking this Epic complete.

- [x] #24468 
- [x] #24469
- [x] #24472
- [x] #24980
- [ ] Finalize CSVFileWriteSchemaTransformProvider

### Issue Priority

Priority: 2

### Issue Component

Component: io-java-files

### LINKED ISSUE #24980: [Task]: Replace CsvIOTestJavaBeans with SchemaAwareJavaBeans
### What needs to happen?

The PR #24806 introduces `org.apache.beam.sdk.io.common.SchemaAwareJavaBeans`.  To work concurrently, PR #24630 redundantly embeds a similar classes inside `org.apache.beam.sdk.io.csv.CsvIOTestJavaBeans`.  When both #24806 and #24630 resolve, then to prevent code duplication, refactor tests in sdks/java/io/csv/src/test to use classes from `SchemaAwareJavaBeans`.

### Issue Priority

Priority: 2 (default / most normal work should be filed as P2)

### Issue Components

- [ ] Component: Python SDK
- [X] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [X] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### PR DESCRIPTION
**Pull request by damondouglas**
This PR closes #24980, and closes #24464 by:
- Replacing CsvIOTestJavaBeans with SchemaAwareJavaBeans
- Finalizing CSVFileWriteSchemaTransformProvider
- Addressing comments from [https://github.com/apache/beam/pull/24630#pullrequestreview-1273656707](https://www.google.com/url?sa=D&q=https%3A%2F%2Fgithub.com%2Fapache%2Fbeam%2Fpull%2F24630%23pullrequestreview-1273656707)

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - ~Update `CHANGES.md` with noteworthy changes.~
 - ~If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).~

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: sdks/java/io/csv/src/main/java/org/apache/beam/sdk/io/csv/CsvIO.java [hunk 1]
```diff
@@ -439,12 +475,16 @@ private static TextIO.Write writeWithCSVFormatHeaderAndComments(
...
+      // which is why we need to manipulate CSVFormat in this way to be compatible
+      // in a TextIO.Write context.
+      CSVFormat withHeaderCommentsRemoved = csvFormat.withHeaderComments();
 
       result.add(
-          withoutHeaderComments
+          withHeaderCommentsRemoved
               // The withSkipHeaderRecord parameter prevents CSVFormat from outputting two copies of
-              // the header.
+              // the header; we already
```
Comment (damondouglas) [inline sdks/java/io/csv/src/main/java/org/apache/beam/sdk/io/csv/CsvIO.java:487]: TODO: @damondouglas Fix type when checks finalize

**Comment by damondouglas** [comment]
Run Whitespace PreCommit

**Comment by damondouglas** [comment]
Seeing:
```
11:56:16 Setting status of ca4de52a57e3e6993952c61864a5f3e34936b6a1 to SUCCESS with url https://ci-beam.apache.org/job/beam_PreCommit_Spotless_Commit/25237/ and message: 'Build finished. '
11:56:16 Using context: Spotless ("Run Spotless PreCommit")
11:56:16 Could not update commit status of the Pull Request on GitHub.
12:14:23 org.kohsuke.github.GHIOException: Ran out of retries for URL: https://api.github.com/repos/apache/beam/statuses/ca4de52a57e3e6993952c61864a5f3e34936b6a1
12:14:23 	at org.kohsuke.github.GitHubClient.sendRequest(GitHubClient.java:407)
12:14:23 	at org.kohsuke.github.GitHubClient.sendRequest(GitHubClient.java:355)
12:14:23 	at org.kohsuke.github.Requester.fetch(Requester.java:76)
12:14:23 	at org.kohsuke.github.GHRepository.createCommitStatus(GHRepository.java:2108)
12:14:23 	at org.jenkinsci.plugins.ghprb.extensions.status.GhprbSimpleStatus.createCommitStatus(GhprbSimpleStatus.java:283)
12:14:23 	at org.jenkinsci.plugins.ghprb.extensions.status.GhprbSimpleStatus.onBuildComplete(GhprbSimpleStatus.java:241)
12:14:23 	at org.jenkinsci.plugins.ghprb.GhprbBuilds.onCompleted(GhprbBuilds.java:205)
12:14:23 	at org.jenkinsci.plugins.ghprb.GhprbBuildListener.onCompleted(GhprbBuildListener.java:28)
12:14:23 	at hudson.model.listeners.RunListener.lambda$fireCompleted$0(RunListener.java:207)
12:14:23 	at jenkins.util.Listeners.lambda$notify$0(Listeners.java:59)
12:14:23 	at jenkins.util.Listeners.notify(Listeners.java:67)
12:14:23 	at hudson.model.listeners.RunListener.fireCompleted(RunListener.java:205)
12:14:23 	at hudson.model.Run.execute(Run.java:1941)
12:14:23 	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:44)
12:14:23 	at hudson.model.ResourceController.execute(ResourceController.java:101)
12:14:23 	at hudson.model.Executor.run(Executor.java:442)
12:14:23 [WS-CLEANUP] Deleting project workspace...
12:14:23 [WS-CLEANUP] Deferred wipeout is used...
12:14:23 [WS-CLEANUP] done
12:14:23 Finished: SUCCESS
```

**Comment by Abacn** [comment]
thanks, will look into it when come back early next week

File: sdks/java/io/csv/build.gradle [hunk 2]
```diff
@@ -31,4 +31,5 @@ dependencies {
     testImplementation project(path: ":sdks:java:core", configuration: "shadowTest")
     testImplementation library.java.junit
     testRuntimeOnly project(path: ":runners:direct-java", configuration: "shadow")
+    testImplementation project(path: ":sdks:java:io:common", configuration: "testRuntimeMigration")
```
Comment (Abacn) [inline sdks/java/io/csv/build.gradle:34]: Notice this "testRuntimeMigration" is actually a workaround migrating to Gradle 7 (#16307) but somehow the TODO gets removed in #16430. 

CC: @lukecwik Is this configuration still a pending cleanup and are new projects able to get rid of this flag? For this PR we can leave it as is for now.

**Review by Abacn — APPROVED** [review]
Thanks, this is pretty complete and looks good to me

File: sdks/java/io/csv/build.gradle (hunk 2 above)
Comment (damondouglas) [reply sdks/java/io/csv/build.gradle:34]: @Abacn Thank you so much for reviewing.  FYI `testImplementation project(path: ":sdks:java:io:common")` causes: `error: package org.apache.beam.sdk.io.common does not exist`

Comment (damondouglas) [reply sdks/java/io/csv/build.gradle:34]: @Abacn per our conversation resolving comment.

**Comment by Abacn** [comment]
The pending Spotless job actually succeeded though status not updated on GitHub UI. Merging for now

**Pull request merged**